### PR TITLE
Advanced Attachment Tokens

### PR DESCRIPTION
**Pull request by ClaysGit**
Covers the following:
- Move Mirage attach functionality to the Explore Helper
  - Classify as a "system modifier", which will be onLoaded before other (planet) modifiers. This ensures attachments can be applied to Mirage.
- Rewrite Stellar Converter functionality in the Explore Helper
  - On attach, the token locks in place hovering over the target planet and gains a "Destroy planet-name" context menu option.
  - When using the "Destroy planet-name" option:
    - Move units + owner tokens to the nearest graveyard
    - Move planet card and attachment tokens and attachment cards all to the Purge bag
    - Remove stats from planet entry
    - Place Stellar Converter token onto the planet and lock it
  - On detach:
    - Reset planet entry using saved state. This will include attachments...unless a save/load happened between destroying and restoring
    - Pull planet card, attachment tokens and attachment cards (that were on the planet) out of the purge bag
- Added support Dimensional Tear tokens
  - Inject a gravity rift into the system that contains their planet
- When loading attachment tokens, DON'T set their orientation. They can be loaded in the wrong order, so trust the orientation they were left at (prevents tech attachments from flipping during load).

Minor stuff:
- Fix name of Vuil'Raith starting tech, Titans Hero (in Unit Helper)
- Add start message to Argent Flight to pick tech, and to Nekro to trash the Dimensional Tear tokens if Vuil'Raith isn't in game
- Mallice tile is tall, so drop attachment tokens from a little bit higher
- System Helper "planetFromPosition" fills missing system data, which may have been lost during a save/load
- Purge bag announces purgings and unpurgings
- Some other tiny changes or bugfixes

### CODE REVIEW

**Comment by ClaysGit** [comment]
Stellar converter when it's first dropped (it is locked in mid-air):
stellar-converter-on-attach

Stellar converter when it's activated (note the chat text):
stellar-converter-on-destroy

After removing the stellar converter:
stellar-converter-on-detach

**Comment by ClaysGit** [comment]
Art issue: The Mirage backside is the same image as the front-side, mirrored. So the text is backwards.

image

**Comment by darrellanderson** [comment]
> Art issue: The Mirage backside is the same image as the front-side, mirrored. So the text is backwards.
> 
> image

Hmm, looks like the modeler didn't unwrap UVs.  OBJ format is pretty easy to parse, can probably write a short script to rewrite UVs to be left/right half of the image instead of blanket reuse -- just need to get a scan of the back side.

File: Objects/TI4_PurgeBag.ttslua [hunk 1]
```diff
@@ -0,0 +1,16 @@
+--- Announces purged objects.
+-- @author Darrell (copied from Deleted Items)
+-- @author Jabberwocky
+--
+
+function onObjectEnterContainer(bag, enterObject)
+    if bag == self then
+        printToAll('PURGING ' .. enterObject.getName())
```
Comment (darrellanderson) [inline Objects/TI4_PurgeBag.ttslua:8]: Good idea.  Maybe I'll set these to an interesting color.

File: Objects/TI4_Helpers/TI4_ExploreHelper.ttslua [hunk 2]
```diff
@@ -933,6 +1007,573 @@ function AttachLib._layoutDecals(decals)
...
+    local function action()
+        attachTokenObject.setLock(true)
+    end
+    local function timeout() end
+    Wait.condition(action, condition, 3, timeout)
+
+    -- Clear context menu
+    attachTokenObject.clearContextMenu()
+
+    -- Inject planet with no stats
```
Comment (darrellanderson) [inline Objects/TI4_Helpers/TI4_ExploreHelper.ttslua:1357]: Is this correct?  Shouldn't it remove the planet?

File: Objects/TI4_Helpers/TI4_ExploreHelper.ttslua [hunk 3]
```diff
@@ -933,6 +1007,573 @@ function AttachLib._layoutDecals(decals)
...
+    end
+
+    -- 'Hover' over target planet (don't disrupt units, tokens, etc.)
+    local planetPosition = systemObject.positionToWorld(planet.position)
+    local tokenPosition = { x = planetPosition.x, y = planetPosition.y + 5, z = planetPosition.z }
+    local collide = false
+    local fast = true
+    attachTokenObject.setPositionSmooth(tokenPosition, collide, fast)
+
+    -- Lock in mid-air as soon as moving finishes.
```
Comment (darrellanderson) [inline Objects/TI4_Helpers/TI4_ExploreHelper.ttslua:1464]: I'm not sure about locking in mid-air, as a general style everything in right click menus should be optional/convenience.

**Review by darrellanderson — APPROVED** [review]
Overall looks great, I love how much simpler the attach code got.  I'll accept and merge this in, but might tweak a few things (following the other review comments).

File: Objects/TI4_Helpers/TI4_ExploreHelper.ttslua (hunk 3 above)
Comment (ClaysGit) [reply Objects/TI4_Helpers/TI4_ExploreHelper.ttslua:1464]: So, doing it this way got around the following:
- It's a big token, and looks goofy placed directly on top of smaller attachments (which puts it at an angle)
- I didn't want dropping it alone to be the trigger to blow stuff up. Too easy to misclick or misplace (particularly in a 3 planet system, which is tight).

I agree that my solution doesn't feel "finished"

File: Objects/TI4_Helpers/TI4_ExploreHelper.ttslua (hunk 2 above)
Comment (ClaysGit) [reply Objects/TI4_Helpers/TI4_ExploreHelper.ttslua:1357]: I think that the number of planets on a system tile can matter for visual reasons. It affects the number of zones in a system, at least. So this felt safer.

File: Objects/TI4_PurgeBag.ttslua (hunk 1 above)
Comment (ClaysGit) [reply Objects/TI4_PurgeBag.ttslua:8]: That would be nice. Should probably also debounce the reporting to reduce spammy messages.

**Pull request merged**